cbauthx/cbauth.go

```diff
@@ -0,0 +1,437 @@
+	a.lock.Lock()
+	a.endpoints = config.Endpoints
+	a.username = config.Username
+	a.password = config.Password
+	if config.ClusterUuid != "" {
+		a.clusterUuid = config.ClusterUuid
+	}
+
+	reconfigureSigCh := a.reconfigureSigCh
+	a.reconfigureSigCh = make(chan struct{})
+
+	a.lock.Unlock()
+
+	close(reconfigureSigCh)
+
+	return nil
+}
+
+func (a *CbAuth) runThread(
+	initialCli *CbAuthClient,
+	initialEndpoint string,
+	initialUsername string,
+	initialPassword string,
+	initialClusterUuid string,
+) {
+	var lock sync.Mutex
+
+	cli := initialCli
+	endpoint := initialEndpoint
+	username := initialUsername
+	password := initialPassword
+	clusterUuid := initialClusterUuid
+
+	go func() {
+		for {
+			<-a.reconfigureSigCh
+			a.logger.Debug("cbauth received reconfigure signal")
+
+			lock.Lock()
+			a.lock.Lock()
+
+			// decide whether we need to close the original client
+			if a.closed ||
+				!slices.Contains(a.endpoints, endpoint) ||
+				username != a.username ||
+				password != a.password ||
+				clusterUuid != a.clusterUuid {
+
+				a.logger.Debug("cbauth is invalidating its current client")
+				cli.Close()
+			}
+
+			a.lock.Unlock()
+			lock.Unlock()
+
+			if a.closed {
+				break
+			}
+		}
+	}()
+
+	for {
+		if initialCli != nil {
+			cli = initialCli
+			initialCli = nil
+		} else {
+			lock.Lock()
+
+			a.lock.Lock()
+			if a.closed {
+				a.lock.Unlock()
+				lock.Unlock()
+				break
+			}
+
+			newEndpoints := a.endpoints
+			newUsername := a.username
+			newPassword := a.password
+			newClusterUuid := a.clusterUuid
+			a.lock.Unlock()
+
+			newCli, newEndpoint, err := a.newClient(
+				context.Background(),
+				newEndpoints,
+				newUsername,
+				newPassword,
+				newClusterUuid)
+			if err != nil {
+				a.logger.Warn("failed to reconnect to cbauth", zap.Error(err))
+
+				a.lock.Lock()
+				a.connectError = err
+				connectWaitCh := a.connectWaitCh
+				a.connectWaitCh = nil
+				a.lock.Unlock()
+
+				if connectWaitCh != nil {
+					close(connectWaitCh)
+				}
+
+				lock.Unlock()
+				continue
+			}
+
+			cli = newCli
+			endpoint = newEndpoint
+			username = newUsername
+			password = newPassword
+			clusterUuid = newClusterUuid
+			lock.Unlock()
+		}
+
+		a.lock.Lock()
+		a.currentClient.Store(cli)
+		a.connectError = nil
+		connectWaitCh := a.connectWaitCh
+		a.connectWaitCh = nil
+		a.lock.Unlock()
+
+		if connectWaitCh != nil {
+			close(connectWaitCh)
+		}
+
+		err := cli.Run()
+		if err != nil {
+			a.logger.Warn("lost connection to cbauth", zap.Error(err))
+		} else {
+			a.logger.Debug("disconnected from cbauth")
+		}
+	}
+}
+
+func (a *CbAuth) CheckUserPass(ctx context.Context, username string, password string) (UserInfo, error) {
+	cli := a.currentClient.Load()
+	info, err := cli.CheckUserPass(ctx, username, password)
+	if err != nil {
+		if errors.Is(err, ErrInvalidAuth) {
+			return UserInfo{}, ErrInvalidAuth
+		}
+
+		a.logger.Debug("failed to check user with cbauth", zap.Error(err))
```

Comment:
Is this technically request level logging?

Reply:
It's at debug level, so should be fine.